exercises/src/main/scala/fpinscala/parallelism/Par.scala

```diff
@@ -17,12 +17,64 @@ object Par {
     def cancel(evenIfRunning: Boolean): Boolean = false 
   }
   
+
+
+
+  case class Map2Future[A,B,C](a: Future[A], b: Future[B], f: (A,B) => C) extends Future[C] {
+    @volatile var cache: Option[C] = None
+    def isDone = cache.isDefined
+    def isCancelled = a.isCancelled || b.isCancelled
+    def cancel(evenIfRunning: Boolean): Boolean = {
+      a.cancel(evenIfRunning) || b.cancel(evenIfRunning)
+    }
+    def get = compute(Long.MaxValue)
+    def get(timeout: Long, units: TimeUnit): C =
+      compute(TimeUnit.NANOSECONDS.convert(timeout, units))
+
+
+    private def compute(timeoutInNanos: Long): C = cache match {
+      case Some(c) => c
+      case None =>
+        val start = System.nanoTime
+        val ar = a.get(timeoutInNanos, TimeUnit.NANOSECONDS)
+        val stop = System.nanoTime
+        val aTime = stop-start
+        val br = b.get(timeoutInNanos - aTime, TimeUnit.NANOSECONDS)
+        val ret = f(ar, br)
+        cache = Some(ret)
+        ret
+    }
+
+  }
+
+  def lazyUnit[A](a: => A): Par[A] = fork(unit(a))
+
+  // 7.4
+  def asyncF[A,B](f: A => B): A => Par[B] = {
+    a => lazyUnit(f(a))
+  }
+
+  // 7.5
+  def sequence[A](ps: List[Par[A]]): Par[List[A]] = ps.foldRight(unit(List()): Par[List[A]])((a1, a2) => map2(a1, a2)((a, la) => a :: la))
```

Comment:
You shouldn't use `List()` to create an empty list, it's generally preferable to use either `Nil`, or in your case `List.empty[A]` would be best to avoid having to cast it when passing it to the fold. I would also rename the variable this way to be easier to understand what they mean. `pa` is a `Par[A]`, `as` is multiple `a`'s, etc.

```
ps.foldRight(unit(List.empty[A]))((pa, pas) => map2(pa, pas)((a, as) => a :: as))
```